keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,1000 @@
+// parameter should be used instead.
+//
+// This function may get deprecated at some point when enough code has been
+// converted to accepting a logger from the caller and direct access to the
+// fallback logger is not needed anymore.
+func TODO() Logger {
+    return fallbackLogger
+}
+
+// Background retrieves the fallback logger. It should not be called before
+// that logger was initialized by the program and not by code that should
+// better receive a logger via its parameters. TODO can be used as a temporary
+// solution for such code.
+func Background() Logger {
+    return fallbackLogger
+}
+```
+
+#### logging helper API
+
+The following helper code will get copied from klog:
+- `ObjectRef`
+- `KMetaData`
+- `KObj`
+- `KRef`
+- `KObjs`
+
+To ensure that a single import is enough in source files, that package will also contain:
+
+```go
+type Logger = logr.Logger
+var NewContext = logr.NewContext
+// plus more as needed...
+```
+
+The implementation of `KObj` and `KObjs` will be either a direct copy of the
+code in klog or an improved version [with performance
+enhancements](https://github.com/kubernetes/kubernetes/issues/106945),
+depending [on benchmark](https://github.com/kubernetes/kubernetes/pull/106594)
+results.
+
+#### Logging in tests
+
+Ginkgo tests do not need to be changed. Because each process only runs one test
+at a time, the global default can be set and local Logger instances can be
+passed around just as in a normal binary.
+
+Unit tests with `go test` require a bit more work. Each test case must
+initialize a `logr/testing` Logger for its own instance of `testing.T`. The
+default log level will be 5, the level
+[recommended](https://github.com/kubernetes/community/blob/9406b4352fe2d5810cb21cc3cb059ce5886de157/contributors/devel/sig-instrumentation/logging.md#logging-conventions)
+for "the steps leading up to errors and warnings" and "for troubleshooting". It
+can be higher than in production binaries because `go test` without `-v` will
+not print the output for test cases that succeeded. It will only be shown for
+failed test cases, and for those the additional log messages may be useful to
+understand the failure. Optionally, logging options can be added to the test
+binary to modify this default log level.
+
+
+API:
+
+```go
+// NewTestContext returns a logger and context for use in a unit test case with
+// the default logging configuration. TestMainWithLogging can be used to enable
+// command line flags that modify those defaults.
+func NewTestContext(t *testing.T) (context.Context, Logger)
+
+// TestMainWithLogging can be used as a test's TestMain implementation.
+// It adds, parses and validates logging command line options before
+// running tests.
+func TestMainWithLogging(m *testing.M)
+```
+
+Example with additional command line options:
+
+```go
+import (
+    "testing"
+    ktesting "k8s.io/klogr/testing"
+)
+
+func TestMain(m *testing.M) {
```

Comment:
Using TestMain to setup test logger seems like an overkill. If I understand it correctly we will need to remember to create `test_main.go` in every directory that has at least one test.I expect that it will be forgotten every time we add a new package. Could we do something more implicit and less error prone?

Reply:
The only alternative that I see is to have an `init` function which unconditionally registers command line flags.

A "convenient" place for that is in "k8s.io/klogr/testing" itself because then it is guaranteed that the command line flags are there whenever a test package uses that logger. But isn't unconditionally registering global flags an anti-pattern that we want to get away from?

We could have it in "k8s.io/klogr/testing/init", similar to "k8s.io/component-base/logs/testinit". Then a single import is enough, but test authors still need to remember to add it.

Reply:
Could we reach out to SIG-testing and get their thoughts?

Reply:
Done, see https://kubernetes.slack.com/archives/C09QZ4DQB/p1643798792450899

If I don't hear otherwise soon, I'll proceed with `k8s.io/klogr/testing/init`. It's a bit simpler than the current approach (albeit doesn't support validation, but that doesn't do anything right now anyway).

Reply:
As with the question about developer recommendations I see this as something that we can still sort out as part of the actual code review.